### PR TITLE
ENH: improve printing of arrays with multi-line reprs

### LINKED ISSUE #15869: array_repr does not behave well for objects with multi-line reprs
One option here would be to implement and/or use `_repr_pretty_` (xref #11270)

Another would be to detect `\n` in the repr, and insert extra indentation before printing.

Current behavior:
```python
class MultiLine:
    def __repr__(self):
        return "Line 1\nLine 2"

np.array([[None, MultiLine()], [MultiLine(), None]])
```
```python
array([[None, Line 1
Line 2],
       [Line 1
Line 2, None]], dtype=object)
```
Ideally the output would be more along the lines of
```
array([[None, Line 1
              Line 2],
       [Line 1
        Line 2, None]], dtype=object)
```
or
```
array([[None,

        Line 1
        Line 2],

       [Line 1
        Line 2,

        None]], dtype=object)
```

### PR DESCRIPTION
**Pull request by panpiort8**
Closes gh-15869. Code is quite inelegant but don't know how to improve it right now.

### CODE REVIEW

File: numpy/core/arrayprint.py [hunk 1]
```diff
@@ -695,7 +695,7 @@ def array2string(a, max_line_width=None, precision=None,
 def _extendLine(s, line, word, line_width, next_line_prefix, legacy):
     needs_wrap = len(line) + len(word) > line_width
     if legacy != '1.13':
-        s# don't wrap lines if it won't help
+        # don't wrap lines if it won't help
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:698]: How the heck did this get here?

File: numpy/core/arrayprint.py [hunk 2]
```diff
@@ -706,6 +706,26 @@ def _extendLine(s, line, word, line_width, next_line_prefix, legacy):
     return s, line
 
 
+def _extendLine_pretty(s, line, word, line_width, next_line_prefix, legacy):
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:709]: Would be good if you could add a docstring

File: numpy/core/arrayprint.py [hunk 3]
```diff
@@ -706,6 +706,26 @@ def _extendLine(s, line, word, line_width, next_line_prefix, legacy):
...
+    line_length = len(line)
+    s, line = _extendLine(
+        s, line, words[0], line_width, next_line_prefix, legacy)
+    for word in words[1::]:
+        s += line.rstrip() + '\n'
+        needs_wrap = line_length + len(word) > line_width
+        if legacy != '1.13':
+            # don't wrap lines if it won't help
+            if line_length <= len(next_line_prefix):
+                needs_wrap = False
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:720]: I suspect if `legacy` is passe we shouldn't do this pretty wrapping at all.

Comment (panpiort8) [reply numpy/core/arrayprint.py]: done

File: numpy/core/arrayprint.py (hunk 2 above)
Comment (panpiort8) [reply numpy/core/arrayprint.py:709]: Did you mean full descripions of all parameters? Or simple note is enough?

File: numpy/core/arrayprint.py [hunk 4]
```diff
@@ -824,6 +845,8 @@ def recurser(index, hanging_indent, curr_width):
         # performance and PyPy friendliness, we break the cycle:
         recurser = None
 
+
+
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:849]:
```suggestion
```
Best not to touch unrelated lines

File: numpy/core/tests/test_arrayprint.py [hunk 5]
```diff
@@ -395,6 +395,37 @@ def test_wide_element(self):
...
+            '  None]]'
+        )
+        assert_equal(
+            np.array_repr(a),
+            'array([[None, Line 1\n'
+            '              Line 2],\n'
+            '       [Line 1\n'
+            '        Line 2, None]], dtype=object)'
+        )
+
```
Comment (eric-wieser) [inline numpy/core/tests/test_arrayprint.py:428]: Since it's the motivating use case, would also be nice to have:
```
    def test_nested_array_repr(self):
        arr = np.empty((2, 2), dtype=object)
        arr[0, 0] = np.eye(2)
        arr[0, 1] = np.eye(3)
        arr[1, 0] = None
        arr[1, 1] = np.ones((3, 1)
        assert_equal(
            repr(arr),
            ...   # put whatever you actually get here.
        )
```

Comment (panpiort8) [reply numpy/core/tests/test_arrayprint.py:428]: Done

File: numpy/core/arrayprint.py (hunk 2 above)
Comment (panpiort8) [reply numpy/core/arrayprint.py:709]: Done

File: numpy/core/arrayprint.py [hunk 6]
```diff
@@ -706,6 +706,27 @@ def _extendLine(s, line, word, line_width, next_line_prefix, legacy):
...
+    line_length = len(line)
+    s, line = _extendLine(
+        s, line, words[0], line_width, next_line_prefix, legacy)
+    for word in words[1::]:
+        s += line.rstrip() + '\n'
+        if line_length + len(word) > line_width \
+                and line_length > len(next_line_prefix):
+            line = next_line_prefix + word
+        else:
+            line = line_length*' ' + word
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:726]: This logic looks a little off to me - my reading is that it will do the wrong thing for an object with a repr like
```
short
long enough to cause the line to wrap
short
```
Could you add a test for that?

Comment (eric-wieser) [reply numpy/core/arrayprint.py:726]: I suspect you actually care about `max(len(word) for word in words))`

Comment (panpiort8) [reply numpy/core/arrayprint.py:726]: I supposed this:
```
[None Line 1
 LoooooooooooooooooongLine 2]
```
would be nicer than this (*):
```
[None Line 1
      LoooooooooooooooooongLin
e 2]
```
(last 'endline' due to end of console). However it doesn't always protect us against this ugly 'endline', so it's indeed nonsense. Using `max(len(word) for word in words))` is by far better option:
```
[None 
 Line 1
 LoooooooooooooooooongLine 2]
```
Anyway after some consideration, I would just remove that if clause, so the result would be:
```
[None Line 1
      LoooooooooooooooooongLin
e 2]
```

Comment (panpiort8) [reply numpy/core/arrayprint.py]: This method is entirely crappy (I hope due to Easter lag). I've changed it a bit

File: numpy/core/tests/test_arrayprint.py [hunk 7]
```diff
@@ -395,6 +395,81 @@ def test_wide_element(self):
...
+        class MultiLineLong:
+            def __repr__(self):
+                return "Line 1\nLooooooooooongestLine2\nLongerLine 3"
+
+        a = np.array([[None, MultiLineLong()], [MultiLineLong(), None]])
+        assert_equal(
+            repr(a),
+            'array([[None, Line 1\n'
+            '              LooooooooooongestLine2\n'
+            '              LongerLine 3          ],\n'
```
Comment (eric-wieser) [inline numpy/core/tests/test_arrayprint.py:438]: Not sure what I think of this suffix padding, let's see whether other reviewers are in favor

File: numpy/core/arrayprint.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/arrayprint.py:698]: Of course it was me... c9052f05ac3a273141f1d222745a420b6b49ed6c

File: numpy/core/arrayprint.py [hunk 8]
```diff
@@ -706,6 +706,34 @@ def _extendLine(s, line, word, line_width, next_line_prefix, legacy):
...
+    Extends line with nicely formatted (possibly multi-line) string ``word``.
+    """
+    words = word.splitlines()
+    if len(words) == 1 or legacy == '1.13':
+        return _extendLine(s, line, word, line_width, next_line_prefix, legacy)
+
+    line_length = len(line)
+    max_word_length = max(len(word) for word in words)
+    if line_length + max_word_length > line_width and \
+            len(line) > len(next_line_prefix):
```
Comment (rossbar) [inline numpy/core/arrayprint.py:720]:
```suggestion
    if (line_length + max_word_length > line_width and
            len(line) > len(next_line_prefix)):
```
Style nit - getting rid of the `\`

File: numpy/core/arrayprint.py [hunk 9]
```diff
@@ -706,6 +706,34 @@ def _extendLine(s, line, word, line_width, next_line_prefix, legacy):
...
+        s += line.rstrip() + '\n'
+        line = next_line_prefix + words[0]
+        indent = next_line_prefix
+    else:
+        line += words[0]
+        indent = line_length*' '
```
Comment (rossbar) [inline numpy/core/arrayprint.py:726]:
```suggestion
    max_word_length = max(len(word) for word in words)
    if len(line) + max_word_length > line_width and \
            len(line) > len(next_line_prefix):
        s += line.rstrip() + '\n'
        line = next_line_prefix + words[0]
        indent = next_line_prefix
    else:
        indent = ' '*len(line)
        line += words[0]
```
It looks like you can eliminate the need for defining the extra `line_length` variable.

**Review by rossbar — COMMENTED** [review]
A couple things stood out, but at first glance this seems to be on the right track for tackling #15869 .

File: numpy/core/tests/test_arrayprint.py (hunk 7 above)
Comment (rossbar) [reply numpy/core/tests/test_arrayprint.py:438]: The suffix padding doesn't conform to any sort of style I'm familiar with, so I'm inclined to be against it. I don't think it does much to improve clarity (at least not in this case). Just my two cents!

Comment (eric-wieser) [reply numpy/core/tests/test_arrayprint.py:438]: Looking again, it's sort of consistent with how we pad floats within a column with spaces to make them all the same width - so I'm +0.5 on keeping it.

File: numpy/core/arrayprint.py [hunk 10]
```diff
@@ -706,6 +706,34 @@ def _extendLine(s, line, word, line_width, next_line_prefix, legacy):
...
+    line_length = len(line)
+    max_word_length = max(len(word) for word in words)
+    if (line_length + max_word_length > line_width and
+            len(line) > len(next_line_prefix)):
+        s += line.rstrip() + '\n'
+        line = next_line_prefix + words[0]
+        indent = next_line_prefix
+    else:
+        line += words[0]
+        indent = line_length*' '
```
Comment (eric-wieser) [inline numpy/core/arrayprint.py:726]: Either eliminate `line_length`

```suggestion
    max_word_length = max(len(word) for word in words)
    if (len(line) + max_word_length > line_width and
            len(line) > len(next_line_prefix)):
        s += line.rstrip() + '\n'
        line = next_line_prefix + words[0]
        indent = next_line_prefix
    else:
        line += words[0]
        indent = len(line)*' '
```

or use it consistently:

```suggestion
    line_length = len(line)
    max_word_length = max(len(word) for word in words)
    if (line_length + max_word_length > line_width and
            line_length > len(next_line_prefix)):
        s += line.rstrip() + '\n'
        line = next_line_prefix + words[0]
        indent = next_line_prefix
    else:
        line += words[0]
        indent = line_length*' '
```

Getting the length of a string is basically free, so I don't think there's much point using the longer `line_length` spelling.

**Review by eric-wieser — APPROVED** [review]

File: numpy/core/tests/test_arrayprint.py (hunk 7 above)
Comment (panpiort8) [reply numpy/core/tests/test_arrayprint.py:438]: I believe in this case it's more clear that first element has multiline representation:
```
array([Line 1
       LooooooooooongestLine2
       LongerLine 3          , None], dtype=object)
```
However I agree that this looks neater:
```
array([Line 1
       LooooooooooongestLine2
       LongerLine 3, None], dtype=object)
```

**Comment by eric-wieser** [comment]
Just needs a release note

File: numpy/core/arrayprint.py (hunk 10 above)
Comment (panpiort8) [reply numpy/core/arrayprint.py]: Sorry, I've messed something with commits. Now it's outdated

Comment (eric-wieser) [reply numpy/core/arrayprint.py]: You've done it already, thanks

File: numpy/core/tests/test_arrayprint.py (hunk 7 above)
Comment (rossbar) [reply numpy/core/tests/test_arrayprint.py:438]: Don't let my stated opinion be a blocker - either one is okay with me, especially given the consistency argument.

**Comment by seberg** [comment]
@panpiort8 could you add a release note snippet? I did not check carefully, but I assume it is otherwise pretty much ready.

Since this is printing related, @ahaldane, Allen (or anyone else) do  you have an opinion on this printing change. This tries to make multiline representations be actually printed on multiple lines with alignment, such as:
```
            array([[array([[1., 0.],
                           [0., 1.]]), array([[1., 0., 0.],
                                              [0., 1., 0.],
                                              [0., 0., 1.]])],
```

**Comment by seberg** [comment]
I am a bit tempted to leave this for 1.20 though, since it may wreak havoc to some doctests.

**Comment by ahaldane** [comment]
The behavior seems good to me, including the whitespace padding, and it seems fine in terms of back-compat doctest issues since it is disabled in "legacy" print mode and only affects multiline elements in object arrays, which I doubt anyone has doctests for. To be safe, waiting for 1.20 seems good.

A big benefit seems to be for smaller ragged nd arrays which are much nicer now:
```python
>>> np.array([rand(3,3), rand(4,4)])  # note, deprecated construction here
array([array([[0.83949546, 0.98664251, 0.68566368],
              [0.82153627, 0.40911358, 0.52227032],
              [0.04588021, 0.50608151, 0.41551733]]),
       array([[0.88704039, 0.45735433, 0.96204133, 0.49452501],
              [0.97484914, 0.16390614, 0.9492255 , 0.29829929],
              [0.33983602, 0.03932543, 0.60143117, 0.33928766],
              [0.37782328, 0.08835779, 0.1864357 , 0.38150337]])],
      dtype=object)
>>> np.array([np.array([np.arange(12), np.arange(10), np.arange(8)]), np.arange(5)]) 
array([array([array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]),
              array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]),
              array([0, 1, 2, 3, 4, 5, 6, 7])], dtype=object)         ,
       array([0, 1, 2, 3, 4])], dtype=object)
```
Note the padding issue discussed in comments above puts one of the commas in the 2nd example pretty far out, but that seems OK to me.

While this particular fix for object arrays seems justified, it seems like we could go on for a long time trying to account for weird strings people's custom objects, so I wouldn't want to add too many more special cases. Besides newlines, the user could have added line-feed, backspace, terminal bell or other control chars, or could also add in commas making the element indistinguishable from two elements. In those cases, I'm inclined to let the user who got themselves into a funny situation make sense of the reprs as they are.

**Comment by eric-wieser** [comment]
Hmm, that comma in the second example is a little irritating.

We could consider taking the approach from ipython.lib.pretty, and wrap the whole row if any element has a multiline repr, allowing us to place that comma more tightly. But that's a lot more work, and a diminishing return.

Leaving till after 1.19.x is branched sounds sane to me.

File: doc/source/release/1.20.0-notes.rst [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+==========================
+NumPy 1.20.0 Release Notes
+==========================
+
+Improvements
+============
+
+Improve `array_repr`
+----------------------------------------------
+Now `array_repr` returns nicely formatted repr for arrays with objects with multi-line reprs.
```
Comment (eric-wieser) [inline doc/source/release/1.20.0-notes.rst:12]: This should go in a newsfragment file by itself. I don't have the docs to hand, but another reviewer not on mobile may remember where they are.

Comment (seberg) [reply doc/source/release/1.20.0-notes.rst:12]: It goes to `doc/release/upcoming_changes`, There is a readme in that directory explaining it. (the header `--` should probably match the title in length, I think sphinx likes to complain about it otherwise).

Comment (mattip) [reply doc/source/release/1.20.0-notes.rst:12]: @panpiort8 could you move the snippet?

**Comment by seberg** [comment]
@panpiort8 as matti mentioned, we would like to merge this, but the snippet needs to be placed in a news fragment file (in `doc/release/upcoming_changes/15997.improvement.rst`)

File: doc/release/upcoming_changes/15997.improvement.rst [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Improve `array_repr`
+--------------------
+Now `array_repr` returns nicely formatted repr for arrays with objects with multi-line reprs.
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/15997.improvement.rst:3]: It's perhaps worth explaining this in simpler terms - this affects `repr(arr)`, not just `np.array_repr(arr)` - so any user of an interactive prompt will see the improvement.

File: doc/release/upcoming_changes/15997.improvement.rst [hunk 13]
```diff
@@ -0,0 +1,4 @@
+Improve `repr(arr)`
+-------------------
+Now `repr(arr)` returns nicely formatted repr for arrays with objects with
+multi-line reprs.
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/15997.improvement.rst:4]:
```suggestion
Object arrays containing multi-line objects have a more readable ``repr``
-------------------------------------------------------------------------
If elements of an object array have a ``repr`` containing new lines, then the
wrapped lines will be aligned by column. Notably, this improves the ``repr`` of
nested arrays::

    >>> np.array([np.eye(2), np.eye(3)], dtype=object)
    array([array([[1., 0.],
                  [0., 1.]]),
           array([[1., 0., 0.],
                  [0., 1., 0.],
                  [0., 0., 1.]])], dtype=object)
```

**Review by eric-wieser — APPROVED** [review]

**Comment by mattip** [comment]
Thanks @panpiort
